File: client/android/app/src/main/java/org/tinymux/titan/data/TimerEngine.kt
package org.tinymux.titan.data

import kotlinx.coroutines.*

data class MudTimer(
    val name: String,
    val command: String,
    val intervalMs: Long,
    var shotsRemaining: Int = -1,
)

class TimerEngine(private val scope: CoroutineScope) {
    private val timers = mutableMapOf<String, Pair<MudTimer, Job>>()
    var onFire: ((timerName: String, command: String) -> Unit)? = null

    fun add(name: String, command: String, intervalMs: Long, shots: Int = -1) {
        remove(name)
        val timer = MudTimer(name, command, intervalMs, shots)
        val job = scope.launch {
            while (isActive) {
                delay(intervalMs)
                onFire?.invoke(name, command)
                if (timer.shotsRemaining > 0) {
                    timer.shotsRemaining--
                    if (timer.shotsRemaining == 0) {
                        timers.remove(name)
                        break
                    }
                }
            }
        }
        timers[name] = timer to job
    }

    fun remove(name: String): Boolean {
        val entry = timers.remove(name)
        entry?.second?.cancel()
        return entry != null
    }

    fun list(): List<MudTimer> = timers.values.map { it.first }

    fun cancelAll() {
        timers.values.forEach { it.second.cancel() }
        timers.clear()
    }
}

File: client/ios/Titan/Model/TimerEngine.swift
import Foundation

// MARK: - Timer Engine

struct MudTimer {
    let name: String
    let command: String
    let intervalSeconds: Double
    var shotsRemaining: Int
}

@MainActor
class TimerEngine {
    private var timers: [String: (MudTimer, Task<Void, Never>)] = [:]
    var onFire: ((String, String) -> Void)?

    func add(name: String, command: String, intervalSeconds: Double, shots: Int = -1) {
        remove(name)
        let initial = MudTimer(name: name, command: command,
                               intervalSeconds: intervalSeconds, shotsRemaining: shots)
        let task = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(intervalSeconds))
                guard !Task.isCancelled else { break }
                guard let self else { break }
                self.onFire?(name, command)
                // The dict is the source of truth: mutate the stored copy so
                // list() reflects the decremented shotsRemaining.
                guard var entry = self.timers[name] else { break }
                if entry.0.shotsRemaining > 0 {
                    entry.0.shotsRemaining -= 1
                    if entry.0.shotsRemaining == 0 {
                        self.timers.removeValue(forKey: name)
                        break
                    } else {
                        self.timers[name] = entry
                    }
                }
            }
        }
        timers[name] = (initial, task)
    }

    @discardableResult
    func remove(_ name: String) -> Bool {
        guard let entry = timers.removeValue(forKey: name) else { return false }
        entry.1.cancel()
        return true
    }

    func list() -> [MudTimer] {
        timers.values.map(\.0)
    }

    func cancelAll() {
        timers.values.forEach { $0.1.cancel() }
        timers.removeAll()
    }
}
